### PR TITLE
feat: add experiment detail actions [DET-3083]

### PR DESCRIPTION
**Pull request by hamidzr**
## Description

depends on the tensorboard launch action PR
## Commentary (optional)

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234

### CODE REVIEW

File: webui/react/src/hooks/useRestApi.ts [hunk 1]
```diff
@@ -45,8 +45,11 @@ type Output<T> = [
 
 const defaultReducer = <T>(state: State<T>, action: Action<T>): State<T> => {
   switch (action.type) {
-    case ActionType.SetData:
-      return { ...state, data: action.value, hasLoaded: true, isLoading: false };
+    case ActionType.SetData: {
+      // TODO unnecessary?
```
Comment (hamidzr) [inline webui/react/src/hooks/useRestApi.ts:49]: this isn't necessary for this PR but @hkang1 do you think this is necessary in general or react is doing something similar under the hood?

File: webui/react/src/components/ExperimentActions.tsx [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+} from 'services/api';
+import { ExperimentDetails, RunState, TBSourceType } from 'types';
+import { openCommand } from 'utils/routes';
+import { cancellableRunStates, killableRunStates, terminalRunStates } from 'utils/types';
+
+import css from './ExperimentActions.module.scss';
+
+interface Props {
+  experiment: ExperimentDetails;
+  updateFn: () => void;
```
Comment (hamidzr) [inline webui/react/src/components/ExperimentActions.tsx:14]: rename to `onSuccess` or `onDone`

**Comment by hamidzr** [comment]
updated this after #836

File: webui/react/src/hooks/useRestApi.ts (hunk 1 above)
Comment (hkang1) [reply webui/react/src/hooks/useRestApi.ts:49]: I wasn't sure either, so I looked into it. Looks like react uses the native JS `Object.is` function to [compare objects](https://reactjs.org/docs/hooks-reference.html#bailing-out-of-a-dispatch). According to the [`Object.is`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/is#Description) it only compares to see if the references are the same for the objects. So in this example, `state.data` is always going to be a different reference than `action.value` (unless you are actually passing in the previous value of `action.value`). But looking at this, it looks like are want to check to make sure the shape and content of the objects are identical so this is NOT redundant code.

Comment (hkang1) [reply webui/react/src/hooks/useRestApi.ts:49]: TL;DR This is great to have :)

File: webui/react/src/components/ExperimentActions.tsx [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+} from 'services/api';
+import { ExperimentDetails, RunState, TBSourceType } from 'types';
+import { openCommand } from 'utils/routes';
+import { cancellableRunStates, killableRunStates, terminalRunStates } from 'utils/types';
+
+import css from './ExperimentActions.module.scss';
+
+interface Props {
+  experiment: ExperimentDetails;
+  finally: () => void; // A callback to trigger after an action is done.
```
Comment (hkang1) [inline webui/react/src/components/ExperimentActions.tsx:14]: **nit:** on callbacks I've been using the format of `on[Action]` so for this maybe we can name it `onComplete`.

File: webui/react/src/components/ExperimentActions.tsx [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+const ExperimentActions: React.FC<Props> = ({ experiment, finally: updateFn }: Props) => {
+
+  interface ButtonLoadingStates {
+    kill: boolean;
+    [RunState.StoppingCanceled]: boolean;
+    archive: boolean;
+    [RunState.Paused]: boolean;
+    [RunState.Active]: boolean;
+    tsb: boolean; // tensorboard
+  }
```
Comment (hkang1) [inline webui/react/src/components/ExperimentActions.tsx:26]: **chore:** move this below `Props` definition

File: webui/react/src/components/ExperimentActions.tsx [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+  const [ buttonStates, setButtonStates ] = useState<ButtonLoadingStates>({
+    [RunState.StoppingCanceled]: false,
+    archive: false,
+    [RunState.Active]: false,
+    kill: false,
+    [RunState.Paused]: false,
+    tsb: false,
+  });
+
+  const killExperimentCB = useCallback(() => {
```
Comment (hkang1) [inline webui/react/src/components/ExperimentActions.tsx:37]: **nit:** rename to `handleKill`

File: webui/react/src/components/ExperimentActions.tsx [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+    setButtonStates(state => ({ ...state, kill: true }));
+    killExperiment({ experimentId: experiment.id })
+      .then(updateFn)
+      .finally(() => setButtonStates(state => ({ ...state, kill: false })));
+  }, [ experiment.id, updateFn ]);
+
+  const archiveCB = useCallback((archive: boolean) =>
```
Comment (hkang1) [inline webui/react/src/components/ExperimentActions.tsx:44]: **nit:** `handleArchive` and maybe move it above `handleKill` for alphabetical ordering?

File: webui/react/src/components/ExperimentActions.tsx [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+
+const ExperimentActions: React.FC<Props> = ({ experiment, finally: updateFn }: Props) => {
+
+  interface ButtonLoadingStates {
+    kill: boolean;
+    [RunState.StoppingCanceled]: boolean;
+    archive: boolean;
+    [RunState.Paused]: boolean;
+    [RunState.Active]: boolean;
+    tsb: boolean; // tensorboard
```
Comment (hkang1) [inline webui/react/src/components/ExperimentActions.tsx:25]: **suggestion:** Create a separate enum just for experiment actions. In this [PR](https://github.com/determined-ai/determined/pull/866/files) I defined something like this:
```
enum Action {
  Activate = 'Activate',
  Archive = 'Archive',
  Cancel = 'Cancel',
  Kill = 'Kill',
  Pause = 'Pause',
  OpenTensorBoard = 'Open Tensorboard',
}
```

Comment (hkang1) [inline webui/react/src/components/ExperimentActions.tsx:25]: **nit:** `tensorboard` instead of `tsb`

File: webui/react/src/components/ExperimentActions.tsx [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+      showIf: (exp): boolean => !experimentWillNeverHaveData(exp),
+    },
+  ];
+
+  return (
+    <ul className={css.base}>
+      {actionButtons
+        .filter(ab => !ab.showIf || ab.showIf(experiment as ExperimentDetails))
+        .map(ab => ab.btn)
+      }
```
Comment (hkang1) [inline webui/react/src/components/ExperimentActions.tsx:163]: **suggestion:** We talk more about this one async if you want. In short, this might have too many layers of abstraction. IMO just having all the JSX buttons listed out here with all the properties set will keep this file shorter and easier to read. For an example, check out this [PR](https://github.com/determined-ai/determined/pull/866/files), specifically `ExperimentList.tsx` lines 224-245. With the current setup, if you need to edit the `Cancel` button click handler you have to find `actionButtons` then look for `cancelButton` then finally `requestExpStateCB`.

File: webui/react/src/utils/data.ts [hunk 9]
```diff
@@ -87,3 +87,8 @@ export const taskStateSorter = (a: State, b: State): number => {
     runStateSortValues[b as RunState] : commandStateSortValues[b as CommandState];
   return aValue - bValue;
 };
+
+export const isEqual = (a: unknown, b: unknown): boolean => {
+  if (a === b) return true;
+  return JSON.stringify(a) === JSON.stringify(b);
+};
```
Comment (hkang1) [inline webui/react/src/utils/data.ts:94]: nice, this is sweet!

File: webui/react/src/ioTypes.ts [hunk 10]
```diff
@@ -136,6 +136,7 @@ const validationHistoryIoType = io.type({
 const trialSummaryIoType = io.type({
   hparams: io.any,
   id: io.number,
+  num_steps: io.number,
```
Comment (hkang1) [inline webui/react/src/ioTypes.ts:139]: **suggestion:** you can leave this to a separate PR but there is a `num_batches: io.number` available now.

**Comment by hkang1** [comment]
**nit:** this can be left for a separate PR
2020-07-14 at 9 59 AM

File: webui/react/src/components/ExperimentActions.tsx (hunk 3 above)
Comment (hamidzr) [reply webui/react/src/components/ExperimentActions.tsx:14]: sg

File: webui/react/src/utils/data.ts (hunk 9 above)
Comment (hamidzr) [reply webui/react/src/utils/data.ts:94]: : ) would `areEqual` make more sense?

File: webui/react/src/components/ExperimentActions.tsx (hunk 3 above)
Comment (hamidzr) [reply webui/react/src/components/ExperimentActions.tsx:14]: with `finally` I'm trying to borrow from the async api pattern and to communicate that this will get called wether the action successfully completes or it fails.

Comment (hamidzr) [reply webui/react/src/components/ExperimentActions.tsx:14]: `onSetteled` ? after https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/allSettled

File: webui/react/src/components/ExperimentActions.tsx (hunk 7 above)
Comment (hamidzr) [reply webui/react/src/components/ExperimentActions.tsx:25]: this is a lot better :+1:

File: webui/react/src/ioTypes.ts (hunk 10 above)
Comment (hamidzr) [reply webui/react/src/ioTypes.ts:139]: I added it in.

File: webui/react/src/components/ExperimentActions.tsx (hunk 7 above)
Comment (hamidzr) [reply webui/react/src/components/ExperimentActions.tsx]: how does this look? 

```
enum Action {
  Archive = 'Archive',
  Kill = 'Kill',
  Tensorboard = 'Tensorboard',
}

type ButtonLoadingStates = Record<Action | RunState, boolean>;
```

File: webui/react/src/components/ExperimentActions.tsx (hunk 8 above)
Comment (hamidzr) [reply webui/react/src/components/ExperimentActions.tsx:159]: hmm let's chat about this.

**Comment by hamidzr** [comment]
thanks for the review I addressed the comments there are two left to chat about

https://github.com/determined-ai/determined/pull/858#discussion_r454448836
and https://github.com/determined-ai/determined/pull/858#issuecomment-658269574

**Comment by hamidzr** [comment]
> **nit:** this can be left for a separate PR
> 2020-07-14 at 9 59 AM

IMO having the kill and cancel buttons solid but the action buttons as outlines looks a bit off. I can make all outlines to match batchations :thinking: .. I'll just keep them as primary till we have the rest of the page in place and see what works best?

File: webui/react/src/components/ExperimentActions.tsx (hunk 3 above)
Comment (hkang1) [reply webui/react/src/components/ExperimentActions.tsx:14]: `onSettled` sounds good to me!

File: webui/react/src/components/ExperimentActions.tsx (hunk 8 above)
Comment (hkang1) [reply webui/react/src/components/ExperimentActions.tsx:150]: We discussed this async and decided the array filter and map still makes sense but decided to move the buttons inline into the `actionButtons` list to reduce vertical file space.

**Comment by hamidzr** [comment]
rebased to resolve conflicts with master.
